### PR TITLE
Handle imperial units in the Bulk Order Management interface

### LINKED ISSUE #8171: Group Buy unit size on Bulk Order Management screen does not match unit on product listing
## Description
image
A large buying club (300 products) has moved to OFN and a large part of their business is putting together bulk orders. The feature doesn't work as they need so they are spending hours every week reconciling orders by hand.

## Expected Behavior

On the bulk order management screen, the group buy unit size should match the unit size already entered for the product. 

## Actual Behaviour

1. We want to make a product available for group buy. For example, French Beans.
image## 

The unit is weight (lb) and the unit of sale is 1 (lb).

2. We edit the variant and select Group Buy Options, setting the Bulk Unit Size to 10 because an order to our supplier must be in quantities of 10lbs of French Beans.
image

3. I place an order, selecting minimum order of 1lb and max of 2lbs. I get an email confirmation showing that I purchased 1lb of French Beans. Great!
image

4. But, when I go to Orders>Bulk Order Management and select this order. I see the following, indicating that the Group Unit Buy Size which is supposed to be 10lbs has been set to 10 grams and my order of 1lb (= 454 grams) is 45.4 times the group buy unit size (10 grams). The current and max fulfilled units don't make sense to me -- how do I know how much remains to be ordered?
image

## Steps to Reproduce

(follow steps as documented above?)

https://openfoodnetwork.net/horg/shop (shop)
https://openfoodnetwork.net/admin/enterprises/horg/edit#!/primary_details (UID 782)

## Animated Gif/Screenshot
(above)

## Workaround

Orders are reconciled manually outside of OFN. Much of their business relies on the bulk/group buy order model, so it takes a long time each week.

## Your Environment

* Version used:
* Browser name and version:
* Operating System and version (desktop or mobile):

### PR DESCRIPTION
**Pull request by jibees**
#### What? Why?

Closes #8171

Imperial units seems to be completly forget for this interface. The aim of this PR is to handle correctly imperial units for the whole BOM interface. 

_NB_
The `BULK UNIT SIZE` in the page `/admin/products/french-beans/group_buy_options` is a bit tricky. 
 - If the unit system used is the metric one, the unit used in this interface should be `g` or `L` (no matter of the product unit, that could be `kg` or `g`)
 - If the unit system used is the imperial one, the unit used in this interface should be the one of the product (ie. `oz` or `lb`)

We should somehow edit the `https://guide.openfoodnetwork.org/basic-features/products-1/group-buy-for-bulk-ordering` page

#### What should we test?
First of all, the _how to reproduce_ from the linked issue #8171 . The interface should now be fixed:

Capture d’écran 2021-12-03 à 15 25 54

I think it could be a good idea to test the BOM interface with the product/variants set with imperial units _before_ and _after_ the PR.

#### Release notes
Handle imperial units in the Bulk Order Management interface
Changelog Category: User facing changes

### CODE REVIEW

File: spec/system/admin/bulk_order_management_spec.rb [hunk 1]
```diff
@@ -790,7 +790,7 @@
       let!(:li1) { create(:line_item_with_shipment, order: o1 ) }
       let!(:li2) { create(:line_item_with_shipment, order: o2 ) }
       let!(:p3) {
-        create(:product_with_option_types, group_buy: true, group_buy_unit_size: 5000,
+        create(:product_with_option_types, group_buy: true, group_buy_unit_size: 5,
```
Comment (jibees) [inline spec/system/admin/bulk_order_management_spec.rb:793]: I have a question here actually.

Once the product unit has been set to `kg` : 
Capture d’écran 2021-12-06 à 10 38 42

If we want to set the bulk unit size to `5kg` should the `Bulk Unit Size` field be 5 (`kg`) or 5000 (`g`)?
Capture d’écran 2021-12-06 à 10 38 24

My understanding is that it should be `5`, but I'm wondering which values are set by our users to make the Bulk Order Management interface works:

 
Capture d’écran 2021-12-06 à 10 42 29

 Please help @openfoodfoundation/train-drivers-product-owners ... 🙏

Comment (RachL) [reply spec/system/admin/bulk_order_management_spec.rb:793]: This is what the user guide says: 

image


So the correct answer is 5000 actually :+1:

**Review by mkllnk — APPROVED** [review]
Great! :crossed_fingers:

**Review by Matt-Yorkley — APPROVED** [review]
Just wondering; is there any case where both imperial and metric are present at the same time? Or is it always just dealing with one at a time?

**Comment by jibees** [comment]
@Matt-Yorkley 
> is there any case where both imperial and metric are present at the same time

You mean: 
 - on the same product? As far as I know, a product can have many variants, but all with the same unit
 - on the same platform? Yes, it is handled one at a time.

**Comment by filipefurtad0** [comment]
Hey @jibees ,

Before this PR the BOM page was indeed showing mixed units.

For `lb` units:

- before

image


- after

image

For `oz` units:

- before

image

- after

image


So two points, so far:
- It seems to me, that `Total Quantity Ordered` must be inferior or equal to the `Max Quantity Ordered `; and this seems to change, after staging the PR
- for `oz` units, these seem to completely disappear after this PR

Please feel free to pick up on these comments @jibees if they are helpful. I still need to check what's going on with the numbers displayed on `Current Fulfilled Units` and `Max Fulfilled Units `, the user guide does not really mention them, but there is a test [here](https://github.com/openfoodfoundation/openfoodnetwork/blob/570eb21469e5db5c7de5543671b89d8ce63ab748/spec/system/admin/bulk_order_management_spec.rb#L815) and the build is green. However I'm still not making sense of the numbers.

**Comment by jibees** [comment]
Thanks @filipefurtad0 for testing this one, and finding some issues that I didn't see ;) 
It seems worthwhile to come back to the `In Dev` step :p

**Comment by filipefurtad0** [comment]
Ok, agreed @jibees  - I'll check if the fields `Current Fulfilled Units` and `Max Fulfilled Units` are behaving correctly on the next iteration :+1:

**Comment by filipefurtad0** [comment]
One question which came about, in discussion with @jibees was the fact that, when inputting the Bulk Size Unit - which is step 2. on "[Actual Behaviour](https://github.com/openfoodfoundation/openfoodnetwork/issues/8171)" - the in put is always in g, as described in the user guide:

https://guide.openfoodnetwork.org/basic-features/products-1/group-buy-for-bulk-ordering#enabling-group-buy-for-a-product

image

So, on this regard, I think we might need to provide some input to help @jibees implement the correct solution:
- should the Bulk Unit Size field (found under `/admin/products/<product_name>/group_buy_options`) also be in `g`, for imperial unit products? If not, then
- should it be always `lb` or `oz`, which one would make more sense?
- also, perhaps it would be useful to display the Bulk Unit Size units in the UI, at this screen:
image

What do you think about this @openfoodfoundation/train-drivers-product-owners ? Pinging @lauriewayne as well :+1:
Thanks for your feedback on this, this is a tricky corner of our app....

**Comment by lauriewayne** [comment]
OMG, the imperial measures are the gift that keeps on giving! I will double check with the big user, but my sense is that grams just confuse Americans (snarky comment removed). I think people would be fine muddling through with using items (like "cases" or "bunches") and pounds, but ounces are pretty much too weird even for us, usually.  

**Adding a note in the UI of what units are being used for Bulk Unit Size is a fantastic idea!**

**Comment by jibees** [comment]
@lauriewayne 🙏 

> which I understand are not supported, am I wrong here?

Should be resolved by https://github.com/openfoodfoundation/openfoodnetwork/issues/8540 I guess (hope? ;))

>  I will double check with the big user

Did you?


What I retain:

 - Remove g for imperial users 
 - ounces (oz) is not even used by users, should be pounds (lb) if possible

Will try to deep dive into this one... It can be a nightmare somehow ;)

**Comment by lauriewayne** [comment]
Thank you @jibees, I apologize for not passing this along! Our big user would love to have ounces too if it is not a complete nightmare. Can you confirm for me whether "items" would be included? In the US we usually do not use grams, but it's possible that Canada, which is more civilized and uses both imperial and metric measures, would like to have those. @tschumilas can you comment?

**Comment by tschumilas** [comment]
I don't know if this is helpful or just makes it more messy ----   In Canada we used to be imperial and then changed to metric a few decades ago.  So the result we use both all the time -- even though we are 'officially' metric.   We DO have OFN situations where an enterprise uses both metric and imperial in the same shop.  (Don't shoot the messenger).  Use cases we have today:  (1) a hub shop has one supplier selling cheese in metric and a different suppleir selling chicken in imperial.    (2)  A farm -  sells own -  but sells chicken in metric and beef in imperial.    It is also quite possible (but not happening today I don't think) that a hubA wants the supplier's product in imperial,  but hub B wants that same supplier's product in metric.    And yes -  anticipating your question -  the same product (chicken)  could have a metric variant (1.5 kg whole bird) and an imperial variant (lbs of ground).   That said -  Canada is likely the only place with such craziness!!   So we will adjust the user guide and  help our users adapt to whatever works and is most expedient on this.   If we need to tell a user,  they have to decide on a unit type for all variants under 1 product -  that will be FINE.    Just getting this working is the top priority.

Also -  on the BOM and 'item' type variants -  I am positive that this use to work!!  I know that when I first had a shop on OFN I was using BOM to order cases of lettuce where the unit type was item.    But I don't know exactly when it stopped,  and the user guide was changed...  I'd being doing my happy dance if BOM worked for 'items' once again -   Really big hugs to you all on this one!

**Comment by jibees** [comment]
Many commits since the last review. Back in `Code review` then. I think it's good now (for both metric and imperial units).

@filipefurtad0 sorry for this long "How to test?" and the back and forth. Don't hesitate to ping/chat with me if something is not clear/irrelevant/or whatever. 

I think there's a lack of automatic testing when imperial system is used, but maybe it could part of be another issue.

File: spec/javascripts/unit/admin/line_items/controllers/line_items_controller_spec.js.coffee [hunk 2]
```diff
@@ -261,24 +261,24 @@ describe "LineItemsCtrl", ->
           scope.formattedValueWithUnitName(1,unitsProduct,unitsVariant)
           expect(Math.round).toHaveBeenCalled()
 
-        it "calls Math.round with the quotient of scale and value, multiplied by 1000", ->
+        it "calls Math.round with the value multiplied by 1000", ->
```
Comment (mkllnk) [inline spec/javascripts/unit/admin/line_items/controllers/line_items_controller_spec.js.coffee:264]: It's good practice to add spec changes to the same commit as the code changes that require them. So you don't have commits with broken specs and you can see why the spec changes are needed.

**Review by mkllnk — APPROVED** [review]
This is wild. I'm not following 100% but the code looks okay.

File: spec/javascripts/unit/admin/line_items/controllers/line_items_controller_spec.js.coffee (hunk 2 above)
Comment (jibees) [reply spec/javascripts/unit/admin/line_items/controllers/line_items_controller_spec.js.coffee:264]: I totally agree, and I usually try to do it as I can. This time, it was wild (as you said) and I missed to link commits to spec changes.

**Comment by Matt-Yorkley** [comment]
> I'm not following 100%

Same :sweat_smile: 

It's not a good sign if both of us struggle to follow it, right?

**Review by Matt-Yorkley — APPROVED** [review]

**Comment by jibees** [comment]
> It's not a good sign if both of us struggle to follow it, right?

Hope this is not related to my code. 
Yes, that's a tricky issue, I've created/updated tests, trying to cover all cases. 
On the other hand, currently, the feature is fully buggy so, I don't think we take to much risk (of course I could have break something else at some point ...).

**Comment by filipefurtad0** [comment]
Awesome @jibees :muscle: 

In the tests below always the bulk unit size used was = 20:

image

The PR does not introduce an unit here, [as mentioned here](https://github.com/openfoodfoundation/openfoodnetwork/pull/8565#issuecomment-1000378159), but I think this can be addressed in a separate issue.


**Sanity test, Kg product - before staging**

image

**Sanity test, Kg product - after staging**

image

- From the pics above we can see that the unit `Group Buy Size` is correctly changed into Kg :+1: 
- for weight items we can see that the `Current Fulfilled Units` and the `Max Fulfilled Units` are displayed in x1000 values. This is not introduced here, and relates to the pre-filled value on the `weight/volume (g)` column:

 
 

So basically, if 100% of the bulk is fulfilled 1000 is displayed instead of 1, as shown in the [user-guide](https://guide.openfoodnetwork.org/basic-features/products-1/group-buy-for-bulk-ordering#adjusting-orders-to-make-complete-batches).

As to the imperial units:

**- lb bulk item**

image

**- oz bulk item**

image

This is correct! The PR fixes the lb / oz units :+1:

**Comment by lauriewayne** [comment]
Wooo hoo hoo!!! Hurrah!
